### DISCUSSION TITLE
How to iterate over list of keys in rnp?

### DISCUSSION THREAD

**Discussion by shemeshg** [rnpgp/rnp#2161] [Q&A]
I want to iterate over the list of keys in my key store and perform some operations on them. However, I am not sure how to do that.

I looked at the generate.c example https://github.com/rnpgp/rnp/blob/main/src/examples/generate.c , but it only shows how to generate and export specific keys, not how to access existing keys in the key store. How can I get the list of keys and their properties (public or private, user, fingerprint, gtip etc.)?

I also tried to include the `rnp_key_store.h` header file that already implement the functionality in modern way, and seems to have some functions for key store manipulation, but I got some compilation errors due to missing inclues. What are the correct steps to include and use this header file?

Here is my CMakeLists.txt file:
```
add_subdirectory(rnp)
add_executable(test 
main.cpp)
target_link_libraries(test
  PRIVATE
    librnp
)
```

main cpp is exactly like `generate.c`

What am I missing? Any help or guidance would be appreciated. Thank you.

**Comment by ni4** [comment]
@shemeshg RNP library is designed to be used via FFI interface, which is available via the header `rnp/rnp.h`.
Our CLI also uses this interface for all the operations, including the key listing (having some C++ wrappers via the `rnpcpp.hpp` header). Please see the function `cli_rnp_t::keys_matching` from the fficli.cpp for the complete example.

**Reply by shemeshg** [reply]
Thanks for the quick replay.

I've modified to code as suggested and now I get 
"Key(s) not found." though (for `rnp` home folder and for `gnupg` homefolder also)
```
macos@macoss-Mac-mini rnpkeys % ./rnpkeys -l
2 keys found
```

main.cpp
```
#include <iostream>
#include <rnp/rnp.h>
#include <fficli.h>

int main(int, char**){
    cli_rnp_t rnp = {};
    rnp_cfg cfg;   

    if (!cli_rnp_setup(&rnp)) {
        return 1;
    }

    cfg.load_defaults();
    //cfg.set_str(CFG_HOMEDIR, "/Users/macos/.gnupg");
    if (!cli_cfg_set_keystore_info(cfg)) {
        ERR_MSG("fatal: cannot set keystore info");
        return 1;
    }
    if (!rnp.init(cfg)) {
        ERR_MSG("fatal: cannot initialise");
        return 1;
    }

    std::vector<rnp_key_handle_t> keys;
    if (!rnp.keys_matching(keys,  "", CLI_SEARCH_DEFAULT)) {
        std::cout<<"Key(s) not found.\n";
    return 1;
    }       
}
```

CmakeLists.txt
```
cmake_minimum_required(VERSION 3.15)
project(shalom VERSION 0.1.0 LANGUAGES C CXX)

include(CTest)
enable_testing()

find_package(JSON-C  )

add_subdirectory(rnp)

add_executable(shalom 
main.cpp
rnp/src/rnp/fficli.cpp
rnp/src/rnpkeys/tui.cpp
rnp/src/rnp/rnpcfg.cpp
)
include_directories(rnp/src/rnp ${JSON-C_INCLUDE_DIR})
target_link_libraries(shalom
  PRIVATE
    librnp    
)

target_include_directories(shalom
  PRIVATE
    "${PROJECT_SOURCE_DIR}/rnp/src"
    "${PROJECT_SOURCE_DIR}/rnp/src/lib"
    "${JSON-C_INCLUDE_DIRS}"
)

set(CPACK_PROJECT_NAME ${PROJECT_NAME})
set(CPACK_PROJECT_VERSION ${PROJECT_VERSION})
include(CPack)

```

**Reply by ni4** [reply]
You are missing key loading call, please see `cli_rnp_t::load_keyrings` code. Also I would recommend not to stick to usage of `cli_rnp_t` class and related methods, as it is not official interface (unlike the `rnp.h`) and could change at any time.

**Comment by shemeshg** [comment]
Thanks,
`rnp.load_keyrings(true);` solved it.

**Reply by shemeshg** [reply]
just finished a simple `hello world`  `rnpgp` git that do not require  adding subfolder or depending indirectly on `cli_rnp_l`
so others who might have the same issue with getting started can quick start with it.

https://github.com/shemeshg/rnpgp-quickstart
